pkg/reconciler/common/releases.go

```diff
@@ -0,0 +1,140 @@
+	return fetch(manifestPath(TargetVersion(instance), instance))
+}
+
+func InstalledManifest(instance v1alpha1.KComponent) (mf.Manifest, error) {
+	switch {
+	case instance.GetStatus().GetVersion() != "":
+		return fetch(manifestPath(instance.GetStatus().GetVersion(), instance))
+	case instance.GetSpec().GetVersion() != "":
+		return fetch(manifestPath(instance.GetSpec().GetVersion(), instance))
+	default:
+		return fetch(manifestPath(latestRelease(instance), instance))
+	}
+}
+
+func fetch(path string) (mf.Manifest, error) {
+	if m, ok := cache[path]; ok {
+		return m, nil
+	}
+	result, err := mf.NewManifest(path)
+	if err == nil {
+		cache[path] = result
+	}
+	return result, err
+}
+
+func componentDir(instance v1alpha1.KComponent) string {
+	koDataDir := os.Getenv(KoEnvKey)
+	switch instance.(type) {
+	case *v1alpha1.KnativeServing:
+		return filepath.Join(koDataDir, "knative-serving")
+	case *v1alpha1.KnativeEventing:
+		return filepath.Join(koDataDir, "knative-eventing")
+	}
+	return ""
+}
+
+func manifestPath(version string, instance v1alpha1.KComponent) string {
+	// TODO: check if file exists and if not, construct URL instead
+	return filepath.Join(componentDir(instance), version)
+}
+
+// sanitizeSemver always adds `v` in front of the version.
+// x.y.z is the standard format we use as the semantic version for Knative. The letter `v` is added for
+// comparison purpose.
+func sanitizeSemver(version string) string {
+	return fmt.Sprintf("v%s", version)
+}
+
+// allReleases returns the all the available release versions
+// available under kodata directory for Knative component.
+func allReleases(instance v1alpha1.KComponent) ([]string, error) {
+	// List all the directories available under kodata
+	pathname := componentDir(instance)
+	fileList, err := ioutil.ReadDir(pathname)
+	if err != nil {
+		return nil, err
+	}
+
+	releaseTags := make([]string, 0, len(fileList))
+	for _, file := range fileList {
+		name := path.Join(pathname, file.Name())
+		pathDirOrFile, err := os.Stat(name)
+		if err != nil {
+			return nil, err
+		}
+		if pathDirOrFile.IsDir() {
+			releaseTags = append(releaseTags, file.Name())
+		}
+	}
+	if len(releaseTags) == 0 {
+		return nil, fmt.Errorf("unable to find any version number for %v", instance)
+	}
+
+	// This function makes sure the versions are sorted in a descending order.
+	sort.Slice(releaseTags, func(i, j int) bool {
+		// The index i is the one after the index j. If i is more recent than j, return true to swap.
+		return semver.Compare(sanitizeSemver(releaseTags[i]), sanitizeSemver(releaseTags[j])) == 1
+	})
+
+	return releaseTags, nil
+}
+
+// latestRelease returns the latest release tag available under kodata directory for Knative component.
+func latestRelease(instance v1alpha1.KComponent) string {
+	vers, err := allReleases(instance)
+	if err != nil {
+		panic(err)
+	}
+	// The versions are in a descending order, so the first one will be the latest version.
```

Comment:
I don't think this is true, semvers sorted lexographically sometimes fails.

For example:
1.1.0
1.10.0
1.2.0

Reply:
That's why we have implement a `sort.Slice` function in `allReleases`, making sure that the result of the list is in the descending order for semantic versions. The semver lib can compare two versions.

Reply:
Aha! thanks, that makes sense. In that case overall this LGTM